capsules/extra/src/ieee802154/driver.rs

```diff
@@ -1042,12 +1088,71 @@ impl<'a, M: device::MacDevice<'a>> device::RxClient for RadioDriver<'a, M> {
+                        //          | data_offset | data_len | mic_len | 15.4 frame |
+                        ///////////////////////////////////////////////////////////////////////////////////////////
+
+                        // 2 bytes for the readwrite buffer metadata (read / write index)
+                        const RING_BUF_METADATA_SIZE: usize = 2;
+
+                        // Confirm the availability of the buffer. A buffer of len 0 is indicative
+                        // of the userprocess not allocating a readwrite buffer. We must also
+                        // confirm that the userprocess correctly formatted the buffer to be of length
+                        // 2 + n * USER_FRAME_MAX_SIZE, where n is the number of user frames that the
+                        // buffer can store.
+                        if rbuf.len() == 0
+                            || (rbuf.len() - RING_BUF_METADATA_SIZE) % USER_FRAME_MAX_SIZE != 0
+                        {
+                            // kernel::debug!("[15.4 Driver] Error - improperly formatted readwrite buffer provided");
+                            return false;
+                        }
+
+                        let mic_len = header.security.map_or(0, |sec| sec.level.mic_len());
+                        let frame_len = data_offset + data_len + mic_len;
+
+                        // Frame length incorporates the PSDU offset, so we must subtract this value
+                        // in addition to incorporating the user frame metadata size to obtain the
+                        // length of the user frame.
+                        let user_frame_len = frame_len + USER_FRAME_METADATA_SIZE - PSDU_OFFSET;
+
+                        let read_index = rbuf[0].get() as usize;
+                        let mut write_index = rbuf[1].get() as usize;
+
+                        let max_pending_rx =
+                            (rbuf.len() - RING_BUF_METADATA_SIZE) / USER_FRAME_MAX_SIZE;
+
+                        // confirm user modifiable metadata is valid (i.e. within bounds of the provided buffer)
+                        if read_index >= max_pending_rx || write_index >= max_pending_rx {
+                            // kernel::debug!("[15.4 driver] Invalid read or write index");
+                            return false;
+                        }
+
+                        let offset = RING_BUF_METADATA_SIZE + (write_index * USER_FRAME_MAX_SIZE);
+
+                        // Copy the entire frame over to userland, preceded by three metadata bytes:
+                        // the data offset, the data length, and the MIC length.
+                        rbuf[(offset + USER_FRAME_METADATA_SIZE)..(offset + user_frame_len)]
+                            .copy_from_slice(&buf[PSDU_OFFSET..frame_len]);
+                        rbuf[offset].set(data_offset as u8);
+                        rbuf[offset + 1].set(data_len as u8);
+                        rbuf[offset + 2].set(mic_len as u8);
```

Comment:
As long as `write_index` and `offset` are derived from content in the buffer, there cannot be any `rbuf[]` in this function.

Reply:
Does confirming `if read_index >= max_pending_rx || write_index >= max_pending_rx` not allow for `rbuf[]`? 

My understanding was that we want to avoid invalid indexing into `rbuf` in the instance when the process provides invalid meta data (read / write index). By checking that the read and write index are valid values, does it not allow us now to index safely into `rbuf`?

(I am assuming that since the capsule is not preempted, this will execute to completion after we verify the parameters and that the userprocess will not have an opportunity to alter the data between when we error check and then perform indexing into `rbuf`. This may however be an inaccurate understanding.)

Reply:
That's fine. But, it does mean we have to be very careful since all checking is on us (like with C) and not with the compiler (which is happy to panic).